core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoticeSchemaGenerator.java

```diff
@@ -0,0 +1,263 @@
+    }
+    return schema;
+  }
+
+  /**
+   * Convenient function to find all notices in given packages and describe their fields.
+   *
+   * <p>The returned map has looks this way:
+   *
+   * <pre>{@code
+   * {
+   *   "AttributionWithoutRoleNotice": {
+   *     "attributionId": String.class,
+   *     "csvRowNumber": Long.class,
+   *   }
+   * }
+   * }</pre>
+   *
+   * @param packages List of packages where notices are declared
+   * @return a map describing all notices in given packages (see above)
+   * @throws IOException
+   */
+  @VisibleForTesting
+  static Map<String, Map<String, Class<?>>> contextFieldsInPackages(List<String> packages)
+      throws IOException {
+    // Return a sorted TreeMap for stable results.
+    Map<String, Map<String, Class<?>>> contextFieldsByNotice = new TreeMap<>();
+    for (Class<Notice> noticeClass : findNoticeSubclasses(packages)) {
+      contextFieldsByNotice.put(noticeClass.getSimpleName(), contextFieldsForNotice(noticeClass));
+    }
+
+    return contextFieldsByNotice;
+  }
+
+  @VisibleForTesting
+  static Map<String, Class<?>> contextFieldsForNotice(Class<? extends Notice> noticeClass) {
+    // Return a sorted TreeMap for stable results.
+    Map<String, Class<?>> fields = new TreeMap<>();
+    for (Field field : noticeClass.getDeclaredFields()) {
+      fields.put(field.getName(), field.getType());
+    }
+    return fields;
+  }
+
+  private static boolean isSubclassOf(Class<?> parent, Class<?> child) {
+    return !child.equals(parent) && parent.isAssignableFrom(child);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void maybeAddNoticeClass(Class<?> clazz, List<Class<Notice>> notices) {
+    if (isSubclassOf(Notice.class, clazz)) {
+      notices.add((Class<Notice>) clazz);
+    }
+  }
+
+  private static boolean belongsToAnyPackage(ClassPath.ClassInfo classInfo, List<String> packages) {
+    for (String packageName : packages) {
+      if (classInfo.getName().startsWith(packageName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean isValidatorClass(Class<?> clazz) {
+    return isSubclassOf(FileValidator.class, clazz)
+        || isSubclassOf(SingleEntityValidator.class, clazz);
+  }
+
+  /**
+   * Finds all subclasses of {@link Notice} that belong to the given packages.
+   *
+   * <p>This function also dives into validator classes that may contain inner notice classes.
+   */
+  @VisibleForTesting
+  static List<Class<Notice>> findNoticeSubclasses(List<String> packages) throws IOException {
+    List<Class<Notice>> notices = new ArrayList<>();
+    for (ClassPath.ClassInfo classInfo :
+        ClassPath.from(ClassLoader.getSystemClassLoader()).getTopLevelClasses()) {
+      if (!belongsToAnyPackage(classInfo, packages)) {
+        continue;
+      }
+      Class<?> clazz = classInfo.load();
+      maybeAddNoticeClass(clazz, notices);
+      if (isValidatorClass(clazz)) {
```

Comment:
Alright, I see how you avoided code duplication here (in comparison to #925).

Reply:
Thanks!